Use `iconv()` for portable encoding conversions

DavisVaughan opened the pull request:
Closes #589 

This PR fixes the failures experienced on Windows due to tests not anticipating that `Encoding(x) <- "unknown"` does different things than it does on Mac/Linux.

We now always use `iconv()` to generate the differently encoded variables required for the tests (except bytes, which requires `Encoding<-` still as far as I can tell).

hadley commented:
Maybe it’d be worthwhile to pull this out into a helper function? Maybe returning a named character vector with all three encodings?

DavisVaughan commented:
A named list where the actual elements were unnamed worked better than a named character vector, due to some weird inconsistencies with `match()` and named character vectors (see the character examples in #599)

hadley commented on tests/testthat/test-dictionary.R:
```diff
@@ -233,12 +207,10 @@ test_that("can take the unique loc of 1d arrays (#461)", {
-
-  latin1 <- iconv(utf8, "UTF-8", "latin1")
+  enc <- encodings()
+  utf8 <- enc$utf8
```
I think it'd be a little nicer to just use `enc$utf` in the tests

DavisVaughan replied:
`enc$latin1` -> `enc$latin` too?

hadley replied:
Yeah, ie don’t assign out the list elements.

DavisVaughan replied:
Ah gotcha, I thought you meant actually change all occurrences of `utf8` to `utf`. I see what you mean now. Done.

hadley commented on tests/testthat/test-dictionary.R:
```diff
@@ -254,29 +226,21 @@ test_that("can use matching functions with strings with different encodings", {
+  latin1 <- enc$latin1
+  bytes <- enc$bytes
 
   expect_equal(vec_match(unknown, unknown), 1L)
```
Maybe it's worth turning this into a nested for loop to test every combination?

DavisVaughan replied:
Did this in a number of places, simplifying many of the tests

DavisVaughan commented:
- Simplified many tests with loops
- Extracted out `encoding_bytes()` as most tests don't need it mixed in with the other encodings, and it makes looping cleaner
